src/dbnode/storage/series/buffer.go

```
@@ -181,287 +167,268 @@ func (b *dbBuffer) Write(
-		// the sake of being able to persist it to disk as a single encoded stream.
-		_, err = bucket.merge()
-		if err != nil {
-			return
-		}
+	checksum := digest.SegmentChecksum(segment)
 
-		// This operation is safe because all of the underlying resources will respect the
-		// lifecycle of the context in one way or another. The "bootstrapped blocks" that
-		// we stream from will mark their internal context as dependent on that of the passed
-		// context, and the Encoder's that we stream from actually perform a data copy and
-		// don't share a reference.
-		streams := bucket.streams(ctx)
-		if len(streams) != 1 {
-			// Should never happen as the call to merge above should result in only a single
-			// stream being present.
-			err = errMoreThanOneStreamAfterMerge
-			return
-		}
+	err = persistFn(id, tags, segment, checksum)
+	if err != nil {
+		return FlushOutcomeErr, err
+	}
 
-		// Direct indexing is safe because canRead guarantees us at least one stream
-		res = streams[0]
-	})
+	if bucket, exists := buckets.writableBucket(WarmWrite); exists {
+		bucket.version = version
+	}
 
-	return res, err
+	return FlushOutcomeFlushedToDisk, nil
 }
 
 func (b *dbBuffer) ReadEncoded(ctx context.Context, start, end time.Time) [][]xio.BlockReader {
 	// TODO(r): pool these results arrays
 	var res [][]xio.BlockReader
-	b.forEachBucketAsc(func(bucket *dbBufferBucket) {
-		if !bucket.canRead() {
-			return
-		}
-		if !start.Before(bucket.start.Add(b.blockSize)) {
-			return
+
+	for _, blockStart := range b.inOrderBlockStarts {
+		if !blockStart.Before(end) || !start.Before(blockStart.Add(b.blockSize)) {
+			continue
 		}
-		if !bucket.start.Before(end) {
-			return
+
+		bv, exists := b.bucketVersionsAt(blockStart)
+		if !exists {
+			// Invariant violated. This means the keys in the bucket map does
+			// not match the sorted keys cache, which should never happen.
+			log := b.opts.InstrumentOptions().Logger()
+			log.Errorf(errBucketMapCacheNotInSync)
+			return nil
```

Should return an error rather than returning nil here.

---

See above comment.

---

Added a response^